multicore_runtime/middleware.py

```diff
@@ -0,0 +1,185 @@
+    os.environ.update(frozen_environment)
+
+    # Add in wsgi_env data, including request headers.
+    os.environ.update(request_environment_for_wsgi_env(wsgi_env))
+
+    # Add in configuration data from env_config.
+    os.environ.update(frozen_env_config_env)
+
+    # Add reserved keys, which draw from wsgi_env as well. These have a very
+    # high priority and so are added nearly last.
+    os.environ.update(reserved_env_keys_for_wsgi_env(wsgi_env))
+
+    # Tweak the environment to hide the service bridge.
+    os.environ.update(get_env_to_hide_service_bridge(wsgi_env))
+
+    return app(wsgi_env, start_response)
+
+  return reset_environment_wrapper
+
+
+def request_environment_for_wsgi_env(wsgi_env):
+  """Get a dict of key-value pairs from wsgi_env that work as env variables.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  # Return all key, value pairs from wsgi_env where value is a string.
+  return {key: value for key, value in wsgi_env.iteritems()
+          if isinstance(value, basestring)}
+
+
+def reserved_env_keys_for_wsgi_env(wsgi_env):
+  """Get a dict for reserved keys based on wsgi_env headers and defaults.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  output = {}
+
+  # Use the default value for a reserved key if the corresponding header is not
+  # set, or if the header exists but its value is blank.
+  for key, default in RESERVED_ENV_KEYS.iteritems():
+    value = wsgi_env.get('HTTP_X_APPENGINE_' + key)
+    output[key] = value or default  # Must be set to a valid value or default.
+
+  return output
+
+
+def get_env_to_hide_service_bridge(wsgi_env):
+  """Generate a dictionary of environment variables to hide the service bridge.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+  output = {}
+
+  # Because the request is coming over the service bridge, the service bridge
+  # host and port that are automatically populated in SERVER_NAME and
+  # SERVER_PORT, respectively.  However, this is an implementation detail that
+  # should not be shown to user code. Instead, we'll rely on the HTTP Host
+  # header to retrieve the hostname used in the original request.  This mimics
+  # the behavior of non-VM App Engine.
+  if 'HTTP_HOST' in wsgi_env:
+    output['SERVER_NAME'] = wsgi_env['HTTP_HOST']
+
+  # Similarly we'll use the HTTPS flag to determine the port used in the
+  # original request.
+  https = wsgi_env.get('HTTP_X_APPENGINE_HTTPS', 'off')
+  if https == 'off':
+    output['SERVER_PORT'] = '80'
+  elif https == 'on':
+    output['SERVER_PORT'] = '443'
+  else:
+    logging.warning('Unrecognized value for HTTPS (%s), won\'t modify '
```

Comment:
nitpick: consider breaking the line after the `(` so that you don't need to break the string on two separate lines.

(Keeping strings together is helpful when you're grepping your code for a specific log line, in this case if you look for `"modify SERVER_PORT"` you wouldn't find this line.)